From 1789ebc59781d837bbb09478e11a2fce0a86a53f Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Fri, 26 Jun 2020 17:48:04 +0200
Subject: [PATCH] PONRTSYS-7705: Set Ingress CTP and BP color marking to
 internal by default

The hardware reset value of the CTP and brigde port color marking is to mark
all frames as green even when they were already marked somewhere else.

This commit changes this behaviour to the internal mode which does not
change the color marking of packets by default.
---
 .../net/ethernet/lantiq/switch-api/gsw_flow_core.c | 38 ++++++++++++++++++++--
 1 file changed, 36 insertions(+), 2 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
index 01676f9107cc..30c4cd314e77 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
@@ -3231,8 +3231,21 @@ static GSW_return_t switch_core_init(void *cdev)
 			ctp_config.nLogicalPortId = j;
 			ctp_config.nSubIfIdGroup = 0;
 			ctp_config.nBridgePortId = j;
+			/* Activate the internal marking mode for all CTPs by
+			 * default. This way the CTP will not change the
+			 * marking, but use the marking from the previous step.
+			 * The original value (GSW_MARKING_ALL_GREEN) would
+			 * overwrite some old marking with green.
+			 */
+			ctp_config.eIngressMarkingMode = GSW_MARKING_INTERNAL_MARKING;
+			ctp_config.eEgressMarkingMode = GSW_MARKING_INTERNAL_MARKING;
+			ctp_config.bEgressMarkingOverrideEnable = true;
+			ctp_config.eEgressMarkingModeOverride = GSW_MARKING_INTERNAL_MARKING;
 			ctp_config.eMask = GSW_CTP_PORT_CONFIG_MASK_BRIDGE_PORT_ID |
-					   GSW_CTP_PORT_CONFIG_BRIDGING_BYPASS;
+					   GSW_CTP_PORT_CONFIG_BRIDGING_BYPASS |
+					   GSW_CTP_PORT_CONFIG_INGRESS_MARKING |
+					   GSW_CTP_PORT_CONFIG_EGRESS_MARKING |
+					   GSW_CTP_PORT_CONFIG_EGRESS_MARKING_OVERRIDE;
 			ret = GSW_CtpPortConfigSet(cdev, &ctp_config);
 
 			if (ret == GSW_statusErr) {
@@ -18703,7 +18716,20 @@ GSW_return_t GSW_CTP_PortAssignmentAlloc(void *cdev, GSW_CTP_portAssignment_t *p
 	}
 
 	memset(&CtpConfig, 0, sizeof(GSW_CTP_portConfig_t));
-	CtpConfig.eMask = GSW_CTP_PORT_CONFIG_MASK_BRIDGE_PORT_ID;
+	CtpConfig.eMask = GSW_CTP_PORT_CONFIG_MASK_BRIDGE_PORT_ID |
+			  GSW_CTP_PORT_CONFIG_INGRESS_MARKING |
+			  GSW_CTP_PORT_CONFIG_EGRESS_MARKING |
+			  GSW_CTP_PORT_CONFIG_EGRESS_MARKING_OVERRIDE;
+	/* Activate the internal marking mode for all CTPs by
+	 * default. This way the CTP will not change the
+	 * marking, but use the marking from the previous step.
+	 * The original value (GSW_MARKING_ALL_GREEN) would
+	 * overwrite some old marking with green.
+	 */
+	CtpConfig.eIngressMarkingMode = GSW_MARKING_INTERNAL_MARKING;
+	CtpConfig.eEgressMarkingMode = GSW_MARKING_INTERNAL_MARKING;
+	CtpConfig.bEgressMarkingOverrideEnable = true;
+	CtpConfig.eEgressMarkingModeOverride = GSW_MARKING_INTERNAL_MARKING;
 	/*Associate the bridge port to the allocated CTP*/
 	CtpConfig.nBridgePortId = parm->nBridgePortId;
 	/*Logical Port and SubInterfaceID will fetch the CTP index
@@ -22401,6 +22427,14 @@ GSW_return_t GSW_BridgePortAlloc(void *cdev, GSW_BRIDGE_portAlloc_t *param)
 	tbl_prog_brdgeport_ingress.val[0] |= 0x7;
 	tbl_prog_brdgeport_egress.val[0] |= 0x7;
 
+	/* Set eIngressMarkingMode to GSW_MARKING_INTERNAL_MARKING.
+	 * Activate the internal marking mode for all bridge ports by default.
+	 * This way the bridge ports will not change the marking, but use the
+	 * marking from the previous step. The original value
+	 * (GSW_MARKING_ALL_GREEN) would overwrite some old marking with green.
+	 */
+	tbl_prog_brdgeport_ingress.val[3] |= 0x1;
+
 	/*Default STP State is GSW_STP_PORT_STATE_FORWARD
 	  it can be changed using GSW_Stp_PortCfgSet*/
 	gswdev->brdgeportconfig_idx[param->nBridgePortId].StpState =
